# Feature Importance Analysis

In this notebook, we analyze the feature importances derived from a Random Forest model trained on gene expression data. The dataset contains various features related to gene expression and chromosomal information.

In [ ]:
#get the combined pckl for the dataframe
import pandas as pd

df = pd.read_pickle('../combined_data.pkl')

display(df.head())

#split X and y
X = df.drop('status', axis=1)
y = df['status']

from sklearn.model_selection import KFold
from sklearn.ensemble import RandomForestClassifier
import numpy as np

kf = KFold(n_splits=5, shuffle=True, random_state=42)
rf = RandomForestClassifier(random_state=42)

feature_importances = []

for train_index, val_index in kf.split(X):
    X_train, X_val = X.iloc[train_index], X.iloc[val_index]
    y_train, y_val = y.iloc[train_index], y.iloc[val_index]
    
    rf.fit(X_train, y_train)
    feature_importances.append(rf.feature_importances_)

# Calculate mean and standard deviation of feature importances
mean_importances = np.mean(feature_importances, axis=0)
std_importances = np.std(feature_importances, axis=0)

status  sum_gene_expr_normalized  mean_gene_expr_normalized  \
0  amplified                  0.000000               0.000000e+00   
1  amplified                  0.006066               4.167977e-07   
2  amplified                  0.000000               0.000000e+00   
3  amplified                  0.002203               1.510508e-07   
4  amplified                  0.000000               0.000000e+00   

   variance_gene_expr_normalized  window100_sum_gene_expr_avg  \
0                   0.000000e+00                     0.000000   
1                   1.770619e-10                     0.003042   
2                   0.000000e+00                     0.002028   
3                   3.707889e-11                     0.002072   
4                   0.000000e+00                     0.001658   

   window100_mean_gene_expr_avg  window100_variance_gene_expr_avg  \
0                  0.000000e+00                      0.000000e+00   
1                  2.083989e-07                      8.853096e-11   
2                  1.389326e-07                      5.902064e-11   
3                  1.419621e-07                      5.353520e-11   
4                  1.135697e-07                      4.282816e-11   

   window10_sum_gene_expr_avg  window10_mean_gene_expr_avg  \
0                    0.000000                 0.000000e+00   
1                    0.003042                 2.083989e-07   
2                    0.002028                 1.389326e-07   
3                    0.002072                 1.419621e-07   
4                    0.001658                 1.135697e-07   

   window10_variance_gene_expr_avg  ...  chromosome_chr22  chromosome_chr3  \
0                     0.000000e+00  ...             False            False   
1                     8.853096e-11  ...             False            False   
2                     5.902064e-11  ...             False            False   
3                     5.353520e-11  ...             False            False   
4                     4.282816e-11  ...             False            False   

   chromosome_chr4  chromosome_chr5  chromosome_chr6  chromosome_chr7  \
0            False            False            False            False   
1            False            False            False            False   
2            False            False            False            False   
3            False            False            False            False   
4            False            False            False            False   

   chromosome_chr8  chromosome_chr9  chromosome_chrX  chromosome_chrY  
0            False            False            False            False  
1            False            False            False            False  
2            False            False            False            False  
3            False            False            False            False  
4            False            False            False            False  

[5 rows x 37 columns]

In [7]:
#show the importance of each feature
importances_df = pd.DataFrame({'feature': X.columns, 'importance': mean_importances, 'std': std_importances})
importances_df = importances_df.sort_values('importance', ascending=False)
display(importances_df)

feature  importance       std
3              window100_sum_gene_expr_avg    0.186346  0.000606
4             window100_mean_gene_expr_avg    0.177921  0.000594
6               window10_sum_gene_expr_avg    0.129314  0.000080
7              window10_mean_gene_expr_avg    0.106499  0.000300
9            dev_gene_expr_normalized_mean    0.102354  0.000107
0                 sum_gene_expr_normalized    0.075541  0.000180
10      dev_mean_gene_expr_normalized_mean    0.053614  0.000218
1                mean_gene_expr_normalized    0.048750  0.000188
31                         chromosome_chr7    0.030213  0.000045
34                         chromosome_chrX    0.017570  0.000107
35                         chromosome_chrY    0.012642  0.000118
13                        chromosome_chr10    0.006617  0.000126
12                         chromosome_chr1    0.003749  0.000060
5         window100_variance_gene_expr_avg    0.003532  0.000037
17                        chromosome_chr14    0.003257  0.000072
30                         chromosome_chr6    0.002992  0.000077
29                         chromosome_chr5    0.002680  0.000037
27                         chromosome_chr3    0.002568  0.000027
22                        chromosome_chr19    0.002412  0.000067
15                        chromosome_chr12    0.002387  0.000036
32                         chromosome_chr8    0.002267  0.000037
33                         chromosome_chr9    0.002211  0.000037
14                        chromosome_chr11    0.002171  0.000019
23                         chromosome_chr2    0.002103  0.000061
11  dev_variance_gene_expr_normalized_mean    0.002015  0.000015
18                        chromosome_chr15    0.001946  0.000020
28                         chromosome_chr4    0.001915  0.000035
16                        chromosome_chr13    0.001875  0.000025
20                        chromosome_chr17    0.001843  0.000063
8          window10_variance_gene_expr_avg    0.001818  0.000010
19                        chromosome_chr16    0.001804  0.000046
24                        chromosome_chr20    0.001527  0.000016
26                        chromosome_chr22    0.001477  0.000018
21                        chromosome_chr18    0.001476  0.000038
25                        chromosome_chr21    0.001325  0.000025
2            variance_gene_expr_normalized    0.001266  0.000005

# Analysis of Feature Importance in Random Forest

This table shows the feature importances derived from a Random Forest model. Below is an interpretation of the key columns:

## Feature Importance Scores
- The `importance` column quantifies the contribution of each feature to the model's predictions. 
  - **Higher values** indicate more significant features.
  - The feature `window100_sum_gene_expr_avg` has the highest importance (0.186346), meaning it plays the most critical role in the model's decision-making.

## Standard Deviation (std)
- The `std` column represents the variability of the importance scores across all trees in the Random Forest.
  - **Lower `std` values** indicate that the importance is consistent across trees.

## Key Insights
1. **Top Features**:
   - Features related to **gene expression averages** (e.g., `window100_sum_gene_expr_avg`, `window100_mean_gene_expr_avg`) are highly influential.
   
2. **Chromosomal Features**:
   - Chromosomal features (e.g., `chromosome_chr7`, `chromosome_chrX`) have relatively lower importance but still contribute slightly to the model.

3. **Low-Importance Features**:
   - Features like `chromosome_chr8` and `chromosome_chr15` have very low importance values, suggesting they contribute minimally to the predictions.
